src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,516 @@
+ *         } catch (Exception e) {     // ok as e is an exception parameter
+ *         }
+ *     }
+ *
+ *     void loops() {
+ *         int j = 12;
+ *         for (int i = 0; j &lt; 11; i++) { // violation, unused local variable 'i'.
+ *         }
+ *         for (int p = 0; j &lt; 11; p++)   // ok
+ *             p /= 2;
+ *     }
+ *
+ *     void lambdas() {
+ *         Predicate&lt;String&gt; obj = (String str) -&gt; { // ok as 'str' is a lambda parameter
+ *             return true;
+ *         };
+ *         obj.test("test");
+ *     }
+ * }
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.local.var}
+ * </li>
+ * </ul>
+ *
+ * @since 9.3
+ */
+@FileStatefulCheck
+public class UnusedLocalVariableCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_LOCAL_VARIABLE = "unused.local.var";
+
+    /**
+     * An array of increment and decrement tokens.
+     */
+    private static final int[] INCREMENT_AND_DECREMENT_TOKENS = {
+        TokenTypes.POST_INC,
+        TokenTypes.POST_DEC,
+        TokenTypes.INC,
+        TokenTypes.DEC,
+    };
+
+    /**
+     * An array of scope tokens.
+     */
+    private static final int[] SCOPES = {
+        TokenTypes.SLIST,
+        TokenTypes.LITERAL_FOR,
+        TokenTypes.OBJBLOCK,
+    };
+
+    /**
+     * An array of unacceptable children of ast of type {@link TokenTypes#DOT}.
+     */
+    private static final int[] UNACCEPTABLE_CHILD_OF_DOT = {
+        TokenTypes.DOT,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.LITERAL_SUPER,
+        TokenTypes.LITERAL_CLASS,
+        TokenTypes.LITERAL_THIS,
+    };
+
+    /**
+     * An array of unacceptable parent of ast of type {@link TokenTypes#IDENT}.
+     */
+    private static final int[] UNACCEPTABLE_PARENT_OF_IDENT = {
+        TokenTypes.VARIABLE_DEF,
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.PATTERN_VARIABLE_DEF,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps VariableDesc objects inside stack of declared variables.
+     */
+    private Deque<VariableDesc> variables;
+
+    @Override
+    public int[] getDefaultTokens() {
+        return new int[] {
+            TokenTypes.DOT,
+            TokenTypes.VARIABLE_DEF,
+            TokenTypes.IDENT,
+            TokenTypes.SLIST,
+            TokenTypes.LITERAL_FOR,
+            TokenTypes.OBJBLOCK,
+        };
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        variables = new ArrayDeque<>();
```

Comment:
https://github.com/checkstyle/checkstyle/pull/10930/files#r754274388
https://github.com/checkstyle/checkstyle/pull/10930/files#r762410744
> we would have to add the survived mutation in the ignore list

Surviving mutation in this instance would be accurate because you have no code saying that we need to clear the list after the first file run. Pitest does not always mean tests can't be created.

I agree this should be changed as other suppression denotes, we should declare the instance as part of the constructor and not need to recreate the class for every file. Field should be cleared on begin tree, not for finishing it. We are only concerned about initializing the fields for the current file.

For creating a test, this is not the first time we have run into this. See https://github.com/checkstyle/checkstyle/blob/1128b5e1c44338b4414d6ec2e55707ac20984399/src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/TestUtil.java#L145 and some examples on how to create a test for this to satisfy pitest.

Reply:
Done